3rdparty/openssl/CMakeLists.txt

```diff
@@ -0,0 +1,1085 @@
+            ${OPENSSL_GENERATED_BUILD}/Makefile
+    COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/configdata.pm
+            ${OPENSSL_GENERATED_BUILD}/configdata.pm
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${OPENSSL_BUILD_INCLUDES}/crypto
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${OPENSSL_BUILD_INCLUDES}/openssl
+    COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/bn_conf.h
+            ${OPENSSL_BUILD_INCLUDES}/crypto/bn_conf.h
+    COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/dso_conf.h
+            ${OPENSSL_BUILD_INCLUDES}/crypto/dso_conf.h
+    COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/opensslconf.h
+            ${OPENSSL_BUILD_INCLUDES}/openssl/opensslconf.h
+    COMMAND ${CMAKE_COMMAND} -E copy ${CMAKE_CURRENT_SOURCE_DIR}/buildinf.h
+            ${OPENSSL_BUILD_INCLUDES}/buildinf.h
+    BUILD_COMMAND ""
+    INSTALL_COMMAND "")
+else ()
+  ExternalProject_Add(
+    openssl_configure
+    SOURCE_DIR ${OPENSSL_DIR}
+    CONFIGURE_COMMAND
+    COMMAND
+      ${OE_PERL} ${OPENSSL_DIR}/Configure ${OPENSSL_CONFIG_OPTIONS}
+      --prefix=${OPENSSL_BUILD} --openssldir=${OPENSSL_BUILD}
+      CC=${OPENSSL_C_COMPILER} CXX=${OPENSSL_CXX_COMPILER}
+    COMMAND ${CMAKE_COMMAND} -E copy Makefile
+            ${OPENSSL_GENERATED_BUILD}/Makefile
+    COMMAND ${CMAKE_COMMAND} -E copy configdata.pm
+            ${OPENSSL_GENERATED_BUILD}/configdata.pm
+    BUILD_COMMAND
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${OPENSSL_BUILD_INCLUDES}/crypto
+    COMMAND ${CMAKE_COMMAND} -E make_directory ${OPENSSL_BUILD_INCLUDES}/openssl
+    COMMAND
+      ${OE_PERL} "-I." -Mconfigdata "${OPENSSL_DIR}/util/dofile.pl" "-oMakefile"
+      ${OPENSSL_DIR}/include/crypto/bn_conf.h.in >
+      ${OPENSSL_BUILD_INCLUDES}/crypto/bn_conf.h
+    COMMAND
+      ${OE_PERL} "-I." -Mconfigdata "${OPENSSL_DIR}/util/dofile.pl" "-oMakefile"
+      ${OPENSSL_DIR}/include/crypto/dso_conf.h.in >
+      ${OPENSSL_BUILD_INCLUDES}/crypto/dso_conf.h
+    COMMAND
+      ${OE_PERL} "-I." -Mconfigdata "${OPENSSL_DIR}/util/dofile.pl" "-oMakefile"
+      ${OPENSSL_DIR}/include/openssl/opensslconf.h.in >
+      ${OPENSSL_BUILD_INCLUDES}/openssl/opensslconf.h
+    # Note: the OpenSSL_version API does not report the complete list of compiler flags unless we provide the list to the mkbuildinf.pl.
+    COMMAND
+      ${OE_PERL} "${OPENSSL_DIR}/util/mkbuildinf.pl" "${OPENSSL_C_COMPILER}"
+      "linux-x86_64" > ${OPENSSL_BUILD_INCLUDES}/buildinf.h
+    INSTALL_COMMAND "")
+endif ()
+
+ExternalProject_Add(
+  openssl_generated
+  SOURCE_DIR ${OPENSSL_DIR}
+  CONFIGURE_COMMAND ""
+  BUILD_COMMAND
+  COMMAND ${CMAKE_COMMAND} -E make_directory ${OPENSSL_BUILD}/crypto/aes
+  COMMAND
+    ${CMAKE_COMMAND} -E env CC=${OPENSSL_C_COMPILER} ${OE_PERL}
+    ${OPENSSL_DIR}/crypto/aes/asm/aesni-mb-x86_64.pl elf
+    ${OPENSSL_BUILD}/crypto/aes/aesni-mb-x86_64.s
```

Comment:
These asm files are consistent on both Linux and Windows so we don't really need to cache them. I'd like to have other people's opinion on this.

Reply:
The reason that I'd prefer checking in the files is because the answers to the following questions are not explained in the PR:
- What benefit do we get from regenerating files that don't change at all each time OE SDK is built? 
- Why are some generated files checked-in whereas some files (these asms) continue to be generated?

Reply:
As a reference, in Intel's SGX port of Open SSL, these files are checked in and are hand-modified to apply LVI mitigations.

https://github.com/intel/intel-sgx-ssl/tree/master/openssl_source/Linux

Reply:
That is only a subset of generated asm files. Intel still use perl to generate the rest of them.

Reply:
> * What benefit do we get from regenerating files that don't change at all each time OE SDK is built?

This makes it easy to synchronize with OpenSSL. OpenSSL never directly have those those asm files in the tree. Instead, it maintains a perl script per asm file.

> * Why are some generated files checked-in whereas some files (these asms) continue to be generated?

This is because of the limitation of perl from git bash. See comments in https://github.com/openenclave/openenclave/blob/ce462122087c180e6624836b04dee698090a6115/3rdparty/openssl/CMakeLists.txt#L68.

Reply:
> That is only a subset of generated asm files. Intel still use perl to generate the rest of them.

At the very least, shouldn't we also be mitigating these specific asm files and checking them in?

Reply:
Agree with anakrish.  Not all the files could be auto mitigated with the perl script. So Intel SGXSSL manually handles some assembly files.

The files for Windows: https://github.com/intel/intel-sgx-ssl/tree/master/openssl_source/Windows
The files for Linux:  https://github.com/intel/intel-sgx-ssl/tree/master/openssl_source/Linux

To support mitigation for the OpenSSL in OE,  suppose the similar things should be done.

Reply:
@lzha101 Thanks for the information. Few question:
- How does Intel decide which assembly file needs to be manually replaced/patched? Does that mean other files work fine (I refer to the full list here: https://github.com/openenclave/openenclave/blob/b71520b53847138e0252efcbc51b265df7a5a548/3rdparty/openssl/CMakeLists.txt#L33)?
- In the file list, there are few files do not have any `lfence` inserted (e.g., `vpaes-x86_64.s` and `x86_64cpuid.s`), do you know why?

Reply:
OpenSSL uses constant-encoded instructions in some cases. The tools that apply LVI mitigations look for mnemonics, not encodings, of instructions that need mitigation. The manually patched assembly files are those that have at least one constant-encoded instruction that's vulnerable to LVI. There is at least one (I might be forgetting something) exception: one (?) perl file is patched in order to mitigate easily identifiable and vulnerable constant-encoded instructions as they appear in the perl file.

Reply:
Thanks for the clarification, Mark! So my current understand on the Intel's modifications are as follows.
- Modifying the perl script to emit `ret` rather than `.byte    0xf3,0xc3`
- Further patching the assembly files generated by the modified perl script. The patched files are those in the links from @lzha101.

Are these accurate?

Reply:
(Sorry for delay Ming-Wei, I tried replying to the auto-email thinking it might result in reply here, but it didn't.)

Yes @Ming-Wei, converting 0xF3, 0xC3 to ret (as you know, it’s actually rep ret, but that’s a different story) is the main change we make in perl.

Reply:
Got it. Thanks!